### PR TITLE
Schema draft

### PR DESCRIPTION
**Pull request by mnsom**
**Feedback from Paulo**
Thank you @mnsom and @ayau8 for your hard work.
First of all I would suggest to follow the YAGNI principle (You Ain't Going to need It). Don't try to cover for all the possible use cases from the beginning - focus on implementing what is apparently needed from the design or business requirements. If the requirements get clarified or change in the future, then you can change the implementation :)

Screenshot 2024-01-10 at 15 23 20
Screenshot 2024-01-10 at 15 23 43
Screenshot 2024-01-10 at 15 23 53

Screenshot 2024-01-10 at 15 24 10
Screenshot 2024-01-10 at 15 24 22

Screenshot 2024-01-10 at 15 24 32

### CODE REVIEW

**Comment by ayau8** [comment]
Thanks Mana for the first draft - think generally the flow looks good. Some comments from my side for consideration:

[Mentee] 
1) Suggest we include another expertises data table as such info is included in Figma. Key info should be domain + years of experiences, and mentee should possibly have several expertises
2) Suggest we include a) "languages", b) "url" for 4 social platforms and c) "resume" within mentee table, wrt Figma 

[Mentor] 
1) Same as mentee related to expertises
2) Would also suggest we include another "experiences" data table on top as such info is included in Figma, and a mentor should possibly have several experiences
3) Suggest we include a) "guidance_from", b) "languages", c) "url" for 4 social platforms and d) "resume" within mentor table, wrt Figma

[Skill] We may want to double check but I think skill for users and project skills to be trained should be different. I'm not sure if we should have a data table particularly - suggest we separate personal skills and projects' skills (perhaps just include it within the respective data tables?)

[CreateMentors] -  “domain” instead of “damain”. But this should be revised in the expertise table as suggested 

[Others] 
1) Should we also double check if sessions (i.e. booking a session) and reviews (i.e. leaving a review) are required in this stage? I saw those in our Figma design
2) Should we check with front-end how we are going to import images? For example, if we're using Cloudinary like in Le Wagon, we probably would need another approach when setting up the models, and possibly don't need to include  "profile_img" in data table?

Happy to discuss further if you have any questions! Thank you

Figma
https://www.figma.com/file/zYBRhihqhRvHDtyEiXtFwW/Design-Team-Space?type=design&node-id=2-3&mode=design&t=O3eNUoIOTmusMCdR-0

**Comment by mnsom** [comment]
Hi @ayau8 , Thank you so much for your review and comments!
I missed some info from the Figma, so your FB was very helpful🙏

I created a updated schema in the following url. Could you please check if my understanding correct and reflect your ideas correctly?
Schema URL: https://kitt.lewagon.com/db/117112

I mostly agree with your suggestions and took them in the updated schema.
However, I'm not sure if we can have multiple data in one column.
e.g. "user_skills" "languages" for the mentee table, "project_skills" for the project table, etc.

I saw an article which said we should avoid having an array in a column before, so I am wondering if we should create different tables for each ("languages" table, "user_skills" table, etc.)
What do you think about this?

--------------------------
[Others]
I agree with your ideas.
Let's ask the following things to the design team and the front-end team.

○To the design team
1. if sessions (i.e. booking a session) and reviews (i.e. leaving a review) are required in this stage
2. confirm if skill for users and project skills to be trained are different

○To the front-end team
1. how we are going to import images
--------------------------

If you have any questions about my comments, please feel free to let me know.
Thank you!

Mana

**Comment by ayau8** [comment]
Thanks @mnsom for the updated schema. Some further questions and comments for consideration - 

[General] Would prefer separate tables for languages / users skills / project skills given maintainability and flexibility in storing information. Suggest going with your ways and we ask for additional opinion from our second reviewer. 

[Projects] Think we can delete skill_id

[Mentees] I'm thinking if we should create field of "mentee_id" under expertises as foreign key and connect mentees primary id to the foreign key. Reason being a mentee has many expertises and an expertise (with different domain and YOE) should belong to a specific mentee. Open to thoughts on this - this will affect how we set up language / project skills / experiences table as well

For the questions, will consolidate mine to your question list below https://docs.google.com/document/d/1aPB58tZEHO8_xGhrTqMu5CGiYFkmK9VuOe3a1Fe8reY/edit

Thank you!

**Comment by mnsom** [comment]
Hi @ayau8 , thank you for your review and FB again!
I reflected your FB and pushed the updated schema. Could you please take a look again?

And thank you for adding your questions in the doc. Do you think we should ask them to the design team and the front-end team before asking the second reviewer to confirm the schema?

Thank you,
Mana

**Comment by ayau8** [comment]
Thanks @mnsom, think the latest version looks good and no additional comments from my side

I agree we should confirm those questions quickly with Design and FE team before asking the second reviewer for review. Suggest we ask tomorrow and see if they can get back by Wed noon (before our next internal meeting)? We should probably note that we might have additional queries re the user journeys going forward but current questions are more focused on how we design the database

If you don't mind, would you mind sending the questions to both teams tomorrow if fine with you too? Sorry I'll be out from morning tomorrow and will not have my laptop around. Many thanks!

**Comment by mnsom** [comment]
Hi @ayau8 , thank you for your review again!

Sure, I will ask the questions on the shared channel now.
My questions in the docs are more related to the user journey, so I will ask only the questions related to the database.
Also, I'm going to write the questions in the message directly instead of sharing the docs so that other team members can answer them easily by phone as well!

Have a nice day :)
Mana

**Review by ayau8 — COMMENTED** [review]
As discussed, we are submitting for review from @achpsident based on the latest information we possess. Thanks in advance!

**Comment by achpsident** [comment]
> As discussed, we are submitting for review from @achpsident based on the latest information we possess. Thanks in advance!

As previously mentioned, please request my review via the 'Reviewers' panel, top-right part of the page. ;)

**Comment by achpsident** [comment]
@mnsom @ayau8 Hi guys, sorry, I said I'd have the PR format review done by yesterday, but it was a very hectic day. When reading these, keep in mind that this is something new to you, and it's only by making mistakes and getting feedback that we internalize things and grow. 

1. The PR branch is based on an older version of master, before the linter PR got merged. Please rebase it on the latest master. (Screenshot from the Fork app I told you guys about, very handy for visualizing this)
Screen Shot 2023-12-19 at 10 04 44

2. The first commit of the branch doesn't need to be there, you can drop it.

After doing these two things, please force push the branch, to make sure origin stays up to date. 

3. For PR messages, please follow the guidelines here https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

Basically, you want to write in the imperative voice, 72 characters maximum, like this:

`Create initial schema draft`

or

`Update DB schema based on feedback`

4. This is a small detail, but its by taking care of the small details that we achieve perfection:

- The PR description was not updated when additional models got added to the schema in the last commit, and the Kitt link also was not updated. Keeping everything up to date makes the reviewer's job easier.

I should have the second part of my review ready by tomorrow, that of the contents of the schema and models. Thanks for your hard work and detailed back-and-forth discussion, it's important to give things a lot of thought, otherwise https://www.youtube.com/watch?v=rR4n-0KYeKQ ;)

**Comment by achpsident** [comment]
@ayau8 @mnsom First off, I want to say that the idea for the Mentorship table was really inspired, a clever way of handling the interdependency of three other tables!

In general, there's a lot I like about the design, but I think we can make things quite a bit simpler. If we look at Mentors and Mentees, they are almost exactly the same, so we can just have a base User class, with Mentors and Mentees being roles/types of that class. (We also probably want an Admin type, with the ability to delete other Users, etc) Since the User will be used for Devise authentication, we will need to generate it using Devise.

Then we can have Projects. Projects will belong to a Mentor(the main project leader, like Paulo) and have many Mentors(like me, Emory, Oscar, etc) and ProjectAreas. (ProjectAreas are things like Backend, Frontend, etc)

Each ProjectArea would have many Mentees, and the Project would have many Mentees through Project Areas.

With the Guidances/Skills/Languages, since they're not customizable by the Mentor/Mentee, rather just being selected from a predetermined list, we don't need to make them separate tables, instead they can just be constant arrays with the various options on the User model itself, like:

GUIDANCES = ["UI Design", "QA", ...]
...

and then we can store them as arrays/json in columns on the User.

With the resume column on Mentors/Mentees, probably it should be bio/biography instead?
Also, should username be something like email_address? 

From the Figma, it looks like Mentors and Mentees can add additional social media urls beyond the preset ones, so instead of hardcoded columns, perhaps we could have a column that stores a Hash/JSON of various social media urls?

Expertises -> years_of_experience (no s). 

As I said, you don't need to agree with all of these suggestions, and it's very possible that I've missed something. Let's consider the schema as our best effort, not as something set in stone.

**Comment by mnsom** [comment]
@achpsident @ayau8 
Thank you for taking a look into our first schema draft and your feedback!

To make sure I understand your feedback correctly, I created a DB schema with this URL(https://kitt.lewagon.com/db/117845).
I created it with a new URL so that we can compare it with the previous one.

○What I agree
I would like to use your idea that we can use a base User class for Mentors and Mentees, and  generate Project/ProjectArea model.

○Questions
1) With the Guidances/Skills/Languages
I saw an article which said we should avoid having an array in a column before, so I tried to avoid it. 
Is it common thing to use array/JSON in a column?
If so, I think your idea that we store them as arrays/json in columns on the Use is better!

2) With the resume column
I saw both "bio" and "resume" sections in the Figma(only for mentor). Can we store bio as text and resume as bytea?
Screenshot 2023-12-26 at 17 44 42


3) With the username
I'm not sure if username is same as email address. It says "Type in alphabet or numbers".
Maybe we should confirm the design team?
Screenshot 2023-12-26 at 17 42 52


4) Additional question
I saw "timezone" section in the Figma. Should we add it to the User model as Array/JSON?
Screenshot 2023-12-26 at 17 46 35

Please feel free to edit the database schema(https://kitt.lewagon.com/db/117845). If I missed something or misunderstood.
Thank you,
Mana

**Comment by ayau8** [comment]
Thanks @achpsident @mnsom for the comments. 

I've further updated the schema based on the following with reference to the Figma:
https://kitt.lewagon.com/db/117845

Users:
- Guidance / Language / Skill: All updated from plural to singular (if decided to store as arrays)
- Social_media_urls -> social_platform (consists of link title + url)
- Included both username (appeared in Figma) and email address (required when sending invite from mentor to mentee for projects)
- If i understand it correctly, Admin should be part of the role? Hence admin_type shouldn't be required
- Added isAvailable, timezone field

Expertises:
- years_of_experience: updated to singular

Project_areas:
- Added project management as proj_man field
Screenshot 2023-12-27 at 4 19 58 PM

Projects:
- Added fend_date and progress field
Screenshot 2023-12-27 at 4 19 38 PM

Reviews:
- Added table based on mentors profile's 
Screenshot 2023-12-27 at 4 17 25 PM

With the Guidances/Skills/Languages, believe it depends on the situation for the use of arrays or separate table. This is to be confirmed with Paulo. 
- If the only purpose is to reflect skills, languages and guidance in the user profile, making use of arrays should serve the purpose and can reduce the tables required for the database
- If eventually it includes complex querying needs, such as filtering or sorting based on these aspects, then it seems the use of separate tables offer greater flexibility compared to using arrays

Thank you!

**Comment by mnsom** [comment]
Hi @ayau8, 
Thank you for your update and answering my questions.
I have checked the schema you updated and it seems good to me.

Thank you!

**Comment by mnsom** [comment]
Hi @Naokimi,
(@ayau8 @achpsident )

Alvin and I created the first schema draft and updated it based on the feedback from Matt.

The below is the URL we would like you to confirm.
schema url: https://kitt.lewagon.com/db/117845
(We haven't reflected this latest schema in the file yet.)

In addition to the db structure, we would like you to confirm if we should create separate tables with the Guidances/Skills/Languages or we should use array columns in User model.

- If the only purpose is to reflect skills, languages and guidance in the user profile, making use of arrays should serve the purpose and can reduce the tables required for the database

- If eventually it includes complex querying needs, such as filtering or sorting based on these aspects, then it seems the use of separate tables offer greater flexibility compared to using arrays

If you have any questions or clarification, please feel free to let us know.

Thank you,
Mana

File: db/migrate/20231208052627_create_mentees.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class CreateMentees < ActiveRecord::Migration[7.0]
+  def change
+    create_table :mentees do |t|
+      t.string :first_name
+      t.string :last_name
```
Comment (Naokimi) [inline db/migrate/20231208052627_create_mentees.rb:5]: dowe want to cover middle names too?

File: db/migrate/20231208052627_create_mentees.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+      t.string :user_name
```
Comment (Naokimi) [inline db/migrate/20231208052627_create_mentees.rb:6]: what is this going to be used for?

File: db/migrate/20231208052627_create_mentees.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+      t.string :location
+      t.string :company
+      t.string :job_title
```
Comment (Naokimi) [inline db/migrate/20231208052627_create_mentees.rb:9]: why is this needed for a mentee?

File: db/migrate/20231208052627_create_mentees.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+      t.text :profile_img
+      t.text :introduction
+      t.text :linkedin_url
+      t.text :behance_url
+      t.text :instagram_url
+      t.text :portfolio_url
```
Comment (Naokimi) [inline db/migrate/20231208052627_create_mentees.rb:15]: I would recommend handling these as a single jsonb column, as the urls might end up being more fluid

File: db/migrate/20231208052627_create_mentees.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+      t.text :resume
```
Comment (Naokimi) [inline db/migrate/20231208052627_create_mentees.rb:16]: is this meant to be as an attachment or written text?

File: db/migrate/20231208052627_create_mentees.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+class CreateMentees < ActiveRecord::Migration[7.0]
+  def change
+    create_table :mentees do |t|
+      t.string :first_name
+      t.string :last_name
+      t.string :user_name
+      t.string :location
+      t.string :company
+      t.string :job_title
+      t.text :profile_img
```
Comment (Naokimi) [inline db/migrate/20231208052627_create_mentees.rb:10]: image_url is generally engough to be as a string, doesn't need to be as text

File: db/migrate/20231208052943_create_mentors.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+class CreateMentors < ActiveRecord::Migration[7.0]
+  def change
+    create_table :mentors do |t|
```
Comment (Naokimi) [inline db/migrate/20231208052943_create_mentors.rb:3]: some points raised at mentee apply to mentor too

File: db/migrate/20231208052943_create_mentors.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+      t.string :first_name
+      t.string :last_name
+      t.string :user_name
+      t.string :location
+      t.text :profile_img
+      t.text :introduction
+      t.string :guidance_from
```
Comment (Naokimi) [inline db/migrate/20231208052943_create_mentors.rb:10]: what is this meant to represent?

File: db/migrate/20231208053131_create_projects.rb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+class CreateProjects < ActiveRecord::Migration[7.0]
+  def change
+    create_table :projects do |t|
+      t.string :name
+      t.text :description
+      t.text :image_url
+      t.date :starting_date
```
Comment (Naokimi) [inline db/migrate/20231208053131_create_projects.rb:7]: would recommend using `start_date` rather than `starting`
also, since we will probably want to cycle through active projects and some might be already finished, it would be good to have an end date as well, despite not being defined from the project's beginning

File: db/migrate/20231208053131_create_projects.rb [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+      t.integer :duration
```
Comment (Naokimi) [inline db/migrate/20231208053131_create_projects.rb:8]: I see the duration now - as mentioned above I would suggest end_date rather than duration, it's easier to work with

File: db/migrate/20231208053131_create_projects.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+class CreateProjects < ActiveRecord::Migration[7.0]
+  def change
+    create_table :projects do |t|
```
Comment (Naokimi) [inline db/migrate/20231208053131_create_projects.rb:3]: One of the mentors will be the Project Owner who will have edit rights to the project. It would be good to have it reflected in the schema

File: db/migrate/20231210065725_create_languages.rb [hunk 12]
```diff
@@ -0,0 +1,11 @@
+class CreateLanguages < ActiveRecord::Migration[7.0]
```
Comment (Naokimi) [inline db/migrate/20231210065725_create_languages.rb:1]: what are these going to be used for? is it regular languages or programming ones?

**Review by Naokimi — COMMENTED** [review]
Thank you @mnsom and @ayau8 for your hard work.
First of all I would suggest to follow the YAGNI principle (You Ain't Going to Need It). Don't try to cover for all the possible use cases from the beginning - focus on implementing what is apparently needed from the design or business requirements. If the requirements get clarified or change in the future, then you can change the implementation :)

**Comment by Naokimi** [comment]
Closed for stale
